crates/oxc_linter/src/rules/eslint/sort_vars.rs

```diff
@@ -0,0 +1,226 @@
+use std::cmp::Ordering;
+
+use oxc_ast::{
+    ast::{BindingPatternKind, VariableDeclarator},
+    AstKind,
+};
+use oxc_diagnostics::OxcDiagnostic;
+use oxc_macros::declare_oxc_lint;
+
+use crate::{context::LintContext, rule::Rule, AstNode};
+
+#[derive(Debug, Default, Clone)]
+pub struct SortVars {
+    ignore_case: bool,
+}
+
+declare_oxc_lint!(
+    /// ### What it does
+    ///
+    /// When declaring multiple variables within the same block, sorting variable names make it
+    /// easier to find necessary variable easier at a later time.
+    ///
+    /// ### Why is this bad?
+    ///
+    /// Unsorted variable declarations can make the code harder to read and maintain.
+    ///
+    /// ### Examples
+    ///
+    /// Examples of **incorrect** code for this rule:
+    /// ```js
+    /// var b, a;
+    /// var a, B, c;
+    /// ```
+    ///
+    /// Examples of **correct** code for this rule:
+    /// ```js
+    /// var a, b, c, d;
+    /// var a, B, d;
+    /// ```
+    SortVars,
+    pedantic,
+    pending
+);
+
+impl Rule for SortVars {
+    fn from_configuration(value: serde_json::Value) -> Self {
+        let ignore_case = value
+            .get(0)
+            .and_then(|v| v.get("ignoreCase"))
+            .and_then(serde_json::Value::as_bool)
+            .unwrap_or(false);
+
+        Self { ignore_case }
+    }
+
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        let AstKind::VariableDeclaration(var_decl) = node.kind() else {
+            return;
+        };
+
+        if var_decl.declarations.len() <= 1 {
+            return;
+        };
+
+        let binding_declarations: Vec<&VariableDeclarator> = var_decl
+            .declarations
+            .iter()
+            .filter(|decl| matches!(decl.id.kind, BindingPatternKind::BindingIdentifier(_)))
+            .collect();
+
+        if binding_declarations.len() <= 1 {
+            return;
+        }
+
+        for idx in 1..binding_declarations.len() {
+            let prev = binding_declarations[idx - 1];
+            let curr = binding_declarations[idx];
+
+            if self.get_sortable_name(prev).cmp(&self.get_sortable_name(curr)) == Ordering::Greater
+            {
+                ctx.diagnostic(
+                    OxcDiagnostic::warn("Variable declarations should be sorted")
+                        .with_label(curr.span),
+                );
+            }
+        }
+    }
+}
+
+impl SortVars {
+    fn get_sortable_name(&self, decl: &VariableDeclarator) -> String {
```

Comment:
Consider returning a Cow so we don't allocate when we don't need to.

Reply:
Do you mean returning this [Cow](https://doc.rust-lang.org/std/borrow/enum.Cow.html)? 

I'm afraid I don't understand what you are proposing, can you give an example of somewhere where this is used in oxlint?

Reply:
```rs
fn get_sortable_name<'a>(&self, decl: &VariableDeclarator<'a>) -> Cow<'a, str> {
      let BindingPatternKind::BindingIdentifier(ident) = &decl.id.kind else {
            unreachable!();
        };

        if self.ignore_case {
            return Cow::Owned(ident.name.to_lowercase());
        }

        Cow::Borrowed(ident.name.as_str())
    }
}
```

better yet, just pass in the thing you want to compare as a parameter and use
`.eq_ignore_ascii_case()`

Reply:
Thanks! We could pass both decl's in theory but we need to test comparison not equality so not sure how `.eq_ignore_ascii_case()` would help

Reply:
Ah, or is that an optimization for the case where it is equal?